src/mono/mono/mini/mini-amd64.c

```diff
@@ -2259,6 +2305,28 @@ mono_arch_emit_call (MonoCompile *cfg, MonoCallInst *call)
 		ainfo = cinfo->args + i;
 
 		in = call->args [i];
+#ifdef MONO_ARCH_HAVE_SWIFTCALL
+		if (mono_method_signature_has_ext_callconv (sig, MONO_EXT_CALLCONV_SWIFTCALL)) {
+			MonoClass *swift_self = mono_class_try_get_swift_self_class ();
+			MonoClass *swift_error = mono_class_try_get_swift_error_class ();
+			MonoClass *klass = mono_class_from_mono_type_internal (sig->params [i]);
+			if (klass) {
+				if (klass == swift_self) {
+					guint32 size = mini_type_stack_size_full (m_class_get_byval_arg (klass), NULL, sig->pinvoke && !sig->marshalling_disabled);
+					g_assert (size == sizeof (target_mgreg_t));
+					ainfo->storage = ArgValuetypeInReg;
+					ainfo->pair_storage [0] = ArgInIReg;
+					ainfo->pair_regs [0] = GINT32_TO_UINT8 (AMD64_R13);
```

Comment:
Why are these done here. I expect the ainfo to be correctly initialized in `get_call_info` ?

Reply:
Yes, it can be moved to `get_call_info`. The similar support should be added for `UnmanagedCallersOnly` which needs to be implemented on the callee side, in the method prolog and epilog. I would rather revert this change and implement this in the same way. We can add two storage variants for Swift types, like ArgSwiftSelf and ArgSwiftSelfOnStack. What do you think?

Reply:
I dont see what `ArgSwiftSelf` is meant to symbolize, or the benefits of its special definition. `ArgInfo` contains information about how to pass an argument, in regs / stack. Swift args are passed also in regs or stack. `get_call_info` should detect if an argument is swift related, and resolve its allocation to `ArgInIReg` or something else, according to what the swift cconv requires. Then the calling code should just work naturally, without explicit intervention to handle swift cconv. (unless it has some problems with allocating args to those special regs)

Reply:
SwiftSelf is considered a standard argument for P/Invoke, where the register allocator always utilizes the dedicated context register. SwiftError is introduced as a new arg storage type ArgSwiftError and passed to the managed-to-native wrapper, but it is not passed to the native function. Here is an example for arm64:
```
managed_to_native_wrapper(swiftself, swifterror)
str &swifterror, [sp, #swift_error_var] // Store the address of swifterror on the stack
ldr R20, swiftself // Swiftself is always passed through the dedicated register
ldr R21, 0x0 // Set SwiftError register to 0
call native function
ldr IP0, [sp, #swift_error_var] // Load &swifterror
str R21, IP0 // Store R21 into swifterror
```